module/zfs/ddt.c

```diff
@@ -1488,145 +1714,514 @@ ddt_repair_table(ddt_t *ddt, zio_t *rio)
+	    (MAX(1, zfs_dedup_log_flush_rounds_max) + 1 - ddt->ddt_flush_pass);
+	if (want_time > remain_time) {
+		/*
+		 * We're behind; try to catch up a bit by doubling the amount
+		 * this round. If we're behind that means we're in a later
+		 * pass and likely have most of the remaining time to
+		 * ourselves. If we're in the last couple of passes, then
+		 * doubling might just take us over the timeout, but probably
+		 * not be much, and it stops us falling behind. If we're
+		 * in the middle passes, there'll be more to do, but it
+		 * might just help us catch up a bit and we'll recalculate on
+		 * the next pass anyway.
+		 */
+		nentries = MIN(ddt->ddt_flush_min, nentries*2);
+	}
+
+	ddt_lightweight_entry_t ddlwe;
+	uint32_t count = 0;
+	while (ddt_log_take_first(ddt, ddt->ddt_log_flushing, &ddlwe)) {
+		ddt_sync_flush_entry(ddt, &ddlwe,
+		    ddlwe.ddlwe_type, ddlwe.ddlwe_class, tx);
+
+		/* End this round if we've synced as much as we need to. */
+		if (++count >= nentries)
+			break;
+	}
+	ddt->ddt_flush_count += count;
+	ddt->ddt_flush_min -= count;
+
+	if (avl_is_empty(&ddt->ddt_log_flushing->ddl_tree)) {
+		/* We emptied it, so truncate on-disk */
+		DDT_KSTAT_ZERO(ddt, dds_log_flushing_entries);
+		ddt_log_truncate(ddt, tx);
+		/* No more passes needed this txg */
+		ddt->ddt_flush_pass = 0;
+	} else {
+		/* More to do next time, save checkpoint */
+		DDT_KSTAT_SUB(ddt, dds_log_flushing_entries, count);
+		ddt_log_checkpoint(ddt, &ddlwe, tx);
+	}
+
+	ddt_sync_update_stats(ddt, tx);
+
+	return (ddt->ddt_flush_pass == 0);
+}
+
+static inline void
+ddt_flush_force_update_txg(ddt_t *ddt, uint64_t txg)
+{
+	/*
+	 * If we're not forcing flush, and not being asked to start, then
+	 * there's nothing more to do.
+	 */
+	if (txg == 0) {
+		/* Update requested, are we currently forcing flush? */
+		if (ddt->ddt_flush_force_txg == 0)
+			return;
+		txg = ddt->ddt_flush_force_txg;
+	}
+
+	/*
+	 * If either of the logs have entries unflushed entries before
+	 * the wanted txg, set the force txg, otherwise clear it.
+	 */
+
+	if ((!avl_is_empty(&ddt->ddt_log_active->ddl_tree) &&
+	    ddt->ddt_log_active->ddl_first_txg <= txg) ||
+	    (!avl_is_empty(&ddt->ddt_log_flushing->ddl_tree) &&
+	    ddt->ddt_log_flushing->ddl_first_txg <= txg)) {
+		ddt->ddt_flush_force_txg = txg;
+		return;
+	}
+
+	/*
+	 * Nothing to flush behind the given txg, so we can clear force flush
+	 * state.
+	 */
+	ddt->ddt_flush_force_txg = 0;
 }
 
 static void
-ddt_sync_table(ddt_t *ddt, dmu_tx_t *tx, uint64_t txg)
+ddt_sync_flush_log(ddt_t *ddt, dmu_tx_t *tx)
 {
-	spa_t *spa = ddt->ddt_spa;
-	ddt_entry_t *dde;
-	void *cookie = NULL;
+	ASSERT(avl_is_empty(&ddt->ddt_tree));
 
-	if (avl_numnodes(&ddt->ddt_tree) == 0)
+	/* Don't do any flushing when the pool is ready to shut down */
+	if (tx->tx_txg > spa_final_dirty_txg(ddt->ddt_spa))
+		return;
+
+	/* Try to flush some. */
+	if (!ddt_sync_flush_log_incremental(ddt, tx))
+		/* More to do next time */
 		return;
```

Comment:
I'll look more into this logic, but my first feeling is that I don't like this multi-pass flushing approach.  The problem is that for each pass ZFS will sync not only all the DDTs, but also all their indirect blocks, then spacemaps and their indirect blocks, MOS and its indirect blocks, etc.  It may increase write inflation, which we are actually trying to reduce here.  In normal operation I'd prefer to see only one DDT sync pass per TXG, and after that only ZFS own convergence, like spacemaps.

Reply:
We were really torn about this during design.

The goal was always to balance our flushing IO against user IO and also IO from other ZFS components. The problem is that there's no central place where we can ask "we've got this much to write, how much can we do?" so we have to do this passive thing of trying a little bit and watching what happens, trying a bit more, until we arrive at something steady that has room for other parts of ZFS to do their thing.

The most obvious alternative I can think of is to only write one chunk per pass, guessing at the ideal size based on how close we've been to the txg timeout each txg, adding a bit more if we're under, and a bit less if we're over. This should converge on an ideal rate? Hard to say. I think it might not respond well to bursts, because its not really getting feedback until the next txg

The current algorithm effectively computes an average time per entry, looks at the remaining time before the txg timeout is reached, and multiplies to get a number of entries to write this txg which is then spread evenly across the remaining passes. This is recalculated each pass, to make sure we're not going over. So possibly we can do somewhere in between by skewing the spread towards the earlier passes.

I was thinking about whether setting `zfs_dedup_log_flush_passes_max=1` gets an approximation, and I think it's not the worst thing. Maybe it's even exactly the "obvious alternative" above, because it will try to adjust to the ideal flow rate based on what's gone on before.

Honestly, I agree - there's probably more/different we can do, whether in the incremental algorithm itself, or as part of a more central IO rate facility or incremental update facility. Right now though, I think the question is just, is this at least good enough? I think so, as it's seen a lot of testing and this hasn't come yet, and I'd like it to see some real use cases that we can adjust towards.

(I promise I tried hard not to just be "no, please don't make me rip all this out before it's even had a chance!" :sweat_smile:)

Reply:
I would not call it a blocker, we can merge this way, but I'll think about possible alternative algorithms.  I am not sure time-based approach used here is the best for all cases.  TXG timeout is supposed to limit transaction time on almost idle pools, while on active ones it is rarely reached (TXGs are usually shorter).  On the other side I worry that on HDD pools with large transactions it may cause log flush starvation.  Plus I am not sure that time you are measuring is really representative, considering all the overheads.  I could criticize it more, but I'll try to be constructive and propose something instead.

Reply:
Thanks, I appreciate this! I'll keep thinking about it too.